### PR TITLE
Di 23275 change data source implementation

### PR DESCRIPTION
**Pull request by michaelatias**

### CODE REVIEW

File: cpi-side/events/helpers/utils.ts [hunk 1]
```diff
@@ -26,3 +26,14 @@ export function groupRelationBlocks<T extends RelationBlock>(arr: T[]){
     return drawFunctionsArray
 }
 
+export function isArrayEquals<T>(arr: T[] | undefined, arr2: T[] | undefined): boolean{
```
Comment (michaelatias) [inline cpi-side/events/helpers/utils.ts:29]: this is going to change there is no need to review that

File: cpi-side/events/helpers/data-builder.ts [hunk 2]
```diff
@@ -40,7 +41,7 @@ export class ListDataBuilder{
     }
 
     private isDataNeedsToChange(): boolean{
-        return Boolean(this.changes.ListKey || this.changes.ViewKey || this.changes.SearchString || this.changes.SmartSearchQuery || this.changes.PageIndex)
+        return Boolean(this.changes.ListKey != this.state.ListKey || this.changes.ViewKey != this.changes.ViewKey || this.changes.SearchString != this.state.SearchString || isArrayEquals(this.changes.SmartSearchQuery, this.state.SmartSearchQuery) || this.changes.PageIndex != this.state.PageIndex)
```
Comment (michaelatias) [inline cpi-side/events/helpers/data-builder.ts:44]: this is going to change there is no need to review that

File: client-side/src/app/resource-list-0.9/helpers/smart-filters/json-regular-filters/ngx-filters.ts [hunk 3]
```diff
@@ -0,0 +1,345 @@
+import { IPepSmartFilterData, IPepSmartFilterDataValue, IPepSmartFilterOperator, IPepSmartFilterOperatorUnit, PepSmartFilterOperatorUnitType } from "@pepperi-addons/ngx-lib/smart-filters";
```
Comment (michaelatias) [inline client-side/src/app/resource-list-0.9/helpers/smart-filters/json-regular-filters/ngx-filters.ts:1]: all the file is going to change there is no need to review that

File: client-side/src/app/resource-list-0.9/helpers/smart-filters/json-regular-filters.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
+import { IPepSmartFilterData } from "@pepperi-addons/ngx-lib/smart-filters";
```
Comment (michaelatias) [inline client-side/src/app/resource-list-0.9/helpers/smart-filters/json-regular-filters.ts:1]: all the file is going to change there is no need to review that

File: client-side/src/app/resource-list-0.9/helpers/smart-filters/json-regular-filter-factory.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import { IPepSmartFilterData } from "@pepperi-addons/ngx-lib/smart-filters"
```
Comment (michaelatias) [inline client-side/src/app/resource-list-0.9/helpers/smart-filters/json-regular-filter-factory.ts:1]: all the file is going to change there is no need to review that

File: client-side/src/app/resource-list-0.9/helpers/state-observer.ts [hunk 6]
```diff
@@ -1,16 +1,12 @@
...
-import { PepMenuItem } from "@pepperi-addons/ngx-lib/menu"
-import { GVButton } from "src/app/generic-viewer/generic-viewer.model"
+import { ReplaySubject } from "rxjs"
 import { IPepListSortingChangeEvent, PepListSelectionType, PepSelectionData } from "@pepperi-addons/ngx-lib/list"
 import { ListState } from "shared"
-import { stat } from "fs"
 
 export class StateObserver{
     //subjects 
-    private $searchString: Subject<string> = new Subject()
```
Comment (roiaharon-pepperi) [inline client-side/src/app/resource-list-0.9/helpers/state-observer.ts:11]: what is the different between Subject to ReplaySubject?

File: client-side/src/app/resource-list-0.9/helpers/view-blocks-adapter.ts [hunk 7]
```diff
@@ -1,7 +1,20 @@
...
+        switch (type){
+            case "Grid":
+                return new GridViewBlockAdapter(viewBlocks)
+            default:
+                throw Error(`view blocks adapter for type ${type} is not implemented yet`)
+        }
+    }
 }
 
 export class GridViewBlockAdapter implements IViewBlocksAdapter{
```
Comment (roiaharon-pepperi) [inline client-side/src/app/resource-list-0.9/helpers/view-blocks-adapter.ts:20]: the GridViewAdapter implementation should not be in the same file as the interface & factory

File: client-side/src/app/resource-list-0.9/metadata.ts [hunk 8]
```diff
@@ -29,4 +33,10 @@ export type SmartSearchInput = {
         Type: "Menu",
         Fields: SmartSearchField[]
     }
-}
\ No newline at end of file
+}
+
+export function capitalizeFirstLetter(s: string){
```
Comment (roiaharon-pepperi) [inline client-side/src/app/resource-list-0.9/metadata.ts:38]: why do you need this function? what purpose does it serves?

File: client-side/src/app/resource-list-0.9/resource-list/resource-list.component.ts [hunk 9]
```diff
@@ -4,34 +4,79 @@ import { IPepGenericListDataSource } from '@pepperi-addons/ngx-composite-lib/gen
...
   ngOnInit(): void {
     this.load()
-    
-    
   }
 
   async load(){
-    const state = new StateManager(undefined, {ListKey: "LIST_KEY"})
-    this.dataSource = new ListDataSource(this.clientEventService, state)
+    const container = await this.clientEventService.emitLoadListEvent(undefined,{ListKey: "LIST_KEY", SearchString: 'aa'}) as Required<ListContainer>
```
Comment (roiaharon-pepperi) [inline client-side/src/app/resource-list-0.9/resource-list/resource-list.component.ts:44]: why are you emitting this event here and not in the state handler after the generic list init?

File: client-side/src/app/resource-list-0.9/resource-list/resource-list.component.ts [hunk 10]
```diff
@@ -4,34 +4,79 @@ import { IPepGenericListDataSource } from '@pepperi-addons/ngx-composite-lib/gen
...
+    .onSelectionTypeChanged(selectionType => this.selectionType = selectionType)
+    .onViewsMenuChanged(viewsMenu => {
+      this.viewsMenu = viewsMenu
+    })
+    .onSelectedViewChanged(key => this.selectedViewKey = key)
+  }
+
+  subscribeToStateChanges(){
+    this.pepperiList.subscribeToStateChanges()
+    .onPageIndexChanged((index => this.pageIndex.next(index)))
```
Comment (roiaharon-pepperi) [inline client-side/src/app/resource-list-0.9/resource-list/resource-list.component.ts:73]: why do you need it here?

**Review by roiaharon-pepperi — CHANGES_REQUESTED** [review]

File: client-side/src/app/resource-list-0.9/helpers/layout-observer.ts [hunk 11]
```diff
@@ -36,49 +41,63 @@ export class LayoutObserver{
         if(data.search != undefined){
             this.$search.next(data.search)
         }
-        if(data.title){
+        if(data.title != undefined){
             this.$title.next(data.title)
         }
-        if(data.selectionType){
+        if(data.selectionType != undefined){
```
[the remark this answers is not in the document]
Comment (michaelatias) [reply client-side/src/app/resource-list-0.9/helpers/layout-observer.ts:47]: there is no need I'll remove that

File: client-side/src/app/resource-list-0.9/helpers/list-data-source.ts [hunk 12]
```diff
@@ -1,186 +1,23 @@
...
-            const viewBlocksAdapter = new GridViewBlockAdapter(listContainer.Layout.View.ViewBlocks.Blocks)
-            this.dataView = viewBlocksAdapter.adapt()
-        }
-    }
-    //will expose the option to observe the changes on the layout by returning the observer as result
-    subscribeToLayoutChanges(): LayoutObserver{
-    subscribeToLayoutChanges(): LayoutObserver{
-        return this.layoutObserver
-    }
+    constructor(private stateChangedHandler: IStateChangedHandler, private isFirstEvent: boolean = true){
```
[the remark this answers is not in the document]
Comment (michaelatias) [reply client-side/src/app/resource-list-0.9/helpers/list-data-source.ts:10]: in order to identify that the datasource was overridden, and if so there is no need to emit another event

File: client-side/src/app/resource-list-0.9/metadata.ts (hunk 8 above)
Comment (michaelatias) [reply client-side/src/app/resource-list-0.9/metadata.ts:38]: thats for the ngx filters to pepperi filters transformer. will be removed after the pr of the pepperi filters will approved.

File: client-side/src/app/resource-list-0.9/helpers/pepperi-list.ts [hunk 13]
```diff
@@ -103,11 +106,20 @@ export class PepperiList implements IStateChangedHandler{
...
         let listContainer: ListContainer = this.listContainer
-        //in case of first init the container already updated
+        const state = this.stateManager.getState()
+
+        //when first loading the list there is no state so we need to emit load list event without building the changes
+        if(!state){
+            listContainer = await this.getListContainer(this.changes)
+        }
+
+        //in case of first init the container already updated
```
Comment (roiaharon-pepperi) [inline client-side/src/app/resource-list-0.9/helpers/pepperi-list.ts:116]: why do I still need the "isFirstEvent" flag?

File: client-side/src/app/resource-list-0.9/helpers/state-manager.ts [hunk 14]
```diff
@@ -20,14 +20,23 @@ export class StateManager{
 
     buildChangesFromPageParams(params: IPepGenericListParams){
         const changes: Partial<ListState> = {}
+        params.fromIndex = 0
```
Comment (roiaharon-pepperi) [inline client-side/src/app/resource-list-0.9/helpers/state-manager.ts:23]: why are you overwriting this? you are breaking your pagination

File: client-side/src/app/resource-list-0.9/helpers/state-manager.ts [hunk 15]
```diff
@@ -37,6 +46,23 @@ export class StateManager{
...
+
+    private getPageIndexAndPageSize(params: IPepGenericListParams){
+        if(params.fromIndex != undefined && params.toIndex != undefined){
+            const pageSize = params.toIndex - params.fromIndex + 1
+            return {
+                pageIndex: Math.floor(((params.toIndex + 1)/ ( (pageSize) || 1))),
+                pageSize: pageSize
+            }
+        }
+        else if(params.pageIndex != undefined){
```
Comment (roiaharon-pepperi) [inline client-side/src/app/resource-list-0.9/helpers/state-manager.ts:58]: you don't need the else, you are ending the function inside the 'if'

File: client-side/src/app/resource-list-0.9/helpers/state-manager.ts [hunk 16]
```diff
@@ -37,6 +46,23 @@ export class StateManager{
         }
         return changes
     }
+
+    private getPageIndexAndPageSize(params: IPepGenericListParams){
+        if(params.fromIndex != undefined && params.toIndex != undefined){
+            const pageSize = params.toIndex - params.fromIndex + 1
+            return {
+                pageIndex: Math.floor(((params.toIndex + 1)/ ( (pageSize) || 1))),
```
Comment (roiaharon-pepperi) [inline client-side/src/app/resource-list-0.9/helpers/state-manager.ts:54]: not sure this is working.. look on what I did on Resource list 0.7 branch

**Review by roiaharon-pepperi — CHANGES_REQUESTED** [review]

**Review by roiaharon-pepperi — APPROVED** [review]

**Pull request merged**